src/lib/gatk.ml

```diff
@@ -93,67 +93,149 @@ module Configuration = struct
-  *)
-let indel_realigner
-    ?(compress=false)
-    ~configuration
-    ~reference_build
-    ~processors
-    ~run_with input_bam ~output_bam =
-  let open KEDSL in
-  let indel_config, target_config = configuration in
-  let name = sprintf "gatk-%s" (Filename.basename output_bam) in
-  let gatk = Machine.get_tool run_with Tool.Default.gatk in
-  let reference_genome = Machine.get_reference_genome run_with reference_build in
-  let fasta = Reference_genome.fasta reference_genome in
-  let intervals_file =
-    Filename.chop_suffix input_bam#product#path ".bam" ^ "-target.intervals"
-  in
-  let sorted_bam =
-    Samtools.sort_bam_if_necessary
-      ~run_with ~processors ~by:`Coordinate input_bam in
-  let make =
-    Machine.run_program run_with ~name
-      Program.(
-        Tool.(init gatk)
-        && sh ("java -jar $GATK_JAR -T RealignerTargetCreator " ^
-               (String.concat ~sep:" " ([
-                    "-R"; fasta#product#path;
-                    "-I"; sorted_bam#product#path;
-                    "-o"; intervals_file;
-                    "-nt"; Int.to_string processors
-                  ] @ Configuration.Realigner_target_creator.render target_config)))
-        && sh ("java -jar $GATK_JAR -T IndelRealigner"
-               ^ (if compress then " " else " -compress 0 ")
-               ^ (String.concat ~sep:" " ([
-                   "-R"; fasta#product#path;
-                   "-I"; sorted_bam#product#path;
-                   "-o"; output_bam;
-                   "-targetIntervals"; intervals_file;
-                 ] @ Configuration.Indel_realigner.render indel_config))))
-  in
-  let sequence_dict = Picard.create_dict ~run_with fasta in
-  workflow_node ~name
-    (bam_file ~sorting:`Coordinate
-       output_bam ~host:Machine.(as_host run_with))
-    ~make
-    ~edges:[
-      depends_on Tool.(ensure gatk);
-      depends_on fasta;
-      depends_on sorted_bam;
-      depends_on (Samtools.index_to_bai ~run_with sorted_bam);
-      (* RealignerTargetCreator wants the `.fai`: *)
-      depends_on (Samtools.faidx ~run_with fasta);
-      depends_on input_bam;
-      depends_on sequence_dict;
-      on_failure_activate (Remove.file ~run_with output_bam);
-      on_failure_activate (Remove.file ~run_with intervals_file);
-    ]
 
+     We want to be able to run the indel-realigner on mutliple bams, so we
+     cannot use the usual `~result_prefix` argument:
+     See the documentation for the `--nWayOut` option:
+     https://www.broadinstitute.org/gatk/guide/tooldocs/org_broadinstitute_gatk_tools_walkers_indels_IndelRealigner.php#--nWayOut
+     See also 
+     http://gatkforums.broadinstitute.org/gatk/discussion/5588/best-practice-for-multi-sample-non-human-indel-realignment
+
+     On top of that we the GADT `_ KEDSL.bam_or_bams` to return have 2 possible
+     return types:
+     bam_file workflow_node or bam_list workflow_node
+  *)
+open Configuration
+let indel_realigner :
+  type a.
```

Comment:
This is not an existential; this to force the “for all `a`, if my input is `a` bam_or_bams, return `a`.”